docs/api-communication-spec.md

```diff
@@ -0,0 +1,885 @@
+    pcap-api ->> bosh-director: Instances (Deployment)?
+    bosh-director ->> pcap-api: Instances (Deployment)
+    par
+        pcap-api ->> pcap-agent1: Capture pcap(eth0, "host 1.2.3.4", 65k)
+        pcap-api ->> pcap-agent2: Capture pcap(eth0, "host 1.2.3.4", 65k)
+        loop
+            pcap-agent1 ->> pcap-api: pcap data
+            pcap-api ->> pcap-cli: pcap data
+            pcap-agent2 ->> pcap-api: pcap data
+            pcap-api ->> pcap-cli: pcap data
+        end
+    end
+    pcap-cli ->> pcap-api: Stop
+    par
+        pcap-api ->> pcap-agent1: Stop
+        pcap-api ->> pcap-agent2: Stop
+        pcap-agent2 ->> pcap-api: OK (pcap-agent2 stopped)
+        pcap-agent1 ->> pcap-api: OK (pcap-agent1 stopped)
+        pcap-api ->> pcap-cli: Message: CAPTURE_STOPPED (pcap-agent1)
+        pcap-api ->> pcap-cli: Message: CAPTURE_STOPPED (pcap-agent2)
+    end
+
+    pcap-api ->>- pcap-cli: OK
+```
+
+CF Case (with cloud controller and NATS options)
+
+```mermaid
+sequenceDiagram
+   pcap-cli ->>+ pcap-api: Token, Capture Request {<br/>CF (AppID, Instances)<br/>pcap ("eth0", "host 1.2.3.4", 65k) }
+    pcap-api ->> cf-uaa: Verify (Token)
+    cf-uaa ->> pcap-api: Token valid
+    alt Target agent identification via Cloud Controller
+        pcap-api ->> cloud-controller: Instances (AppID, Instances)?
+        cloud-controller ->> pcap-api: Instances (AppID, Instances)
+    else Target agent registry via NATS
+        note right of pcap-api: Look up targets in agent registry
+        pcap-api ->> pcap-api: 
+    end
+    par
+        pcap-api ->> pcap-agent1: Capture pcap(eth0, "host 1.2.3.4", 65k)
+        pcap-api ->> pcap-agent2: Capture pcap(eth0, "host 1.2.3.4", 65k)
+        loop
+            pcap-agent1 ->> pcap-api: pcap data
+            pcap-api ->> pcap-cli: pcap data
+            pcap-agent2 ->> pcap-api: pcap data
+            pcap-api ->> pcap-cli: pcap data
+        end
+    end
+    pcap-cli ->> pcap-api: Stop
+    note over pcap-cli, pcap-api: Client requests a graceful stop
+    par
+        pcap-api ->> pcap-agent1: Stop
+        pcap-api ->> pcap-agent2: Stop
+        pcap-agent2 ->> pcap-api: OK (pcap-agent2 stopped)
+        pcap-agent1 ->> pcap-api: OK (pcap-agent1 stopped)
+        pcap-api ->> pcap-cli: Message: CAPTURE_STOPPED (pcap-agent1)
+        pcap-api ->> pcap-cli: Message: CAPTURE_STOPPED (pcap-agent2)
+    end
+
+    pcap-api ->>- pcap-cli: OK
+
+```
+
+### Authentication Failures
+
+On authentication failures, the message `UNAUTHORIZED` is sent back. The message detail explains the reason for rejection, e.g. signature check failed, token expired, etc.
+
+```mermaid
+sequenceDiagram
+    pcap-cli ->>+ pcap-api: Token, Capture Request {<br/>CF (App, Instances)<br/>pcap ("eth0", "host 1.2.3.4", 65k) }
+    pcap-api ->> cf-uaa: Verify (Token)
+    cf-uaa ->> pcap-api: Token invalid
+    
+    pcap-api ->>- pcap-cli: UNAUTHENTICATED (Reason, e.g. token expired)
```

Comment:
Above it says `UNAUTHORIZED`, which one is it? :smile:

Reply:
can be `UNAUTHENTICATED` or `PERMISSION_DENIED`, nothing else.
https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Reply:
Then please fix the occurrence above